File: other/EducationalDPContest/N/N.py
import sys
sys.setrecursionlimit(1<<31-1)
N=int(input())
A=list(map(int,input().split()))
acc=[0 for _ in range(N+1)]
for i in range(N):
    acc[i+1]=acc[i]+A[i]
dp=[[0 for _ in range(N+1)] for _ in range(N)]
def f(l,r):
    if dp[l][r]>0 or r==l+1:
        return dp[l][r]
    m=float("inf")
    for i in range(l+1,r):
        m=min(m,f(l,i)+f(i,r)+acc[r]-acc[l])
    dp[l][r]=m
    return m
print(f(0,N))

File: other/EducationalDPContest/N/N.cpp
#include <bits/stdc++.h>
using namespace std;
using ll = long long;

int main()
{
	ll N;
	cin >> N;
	vector<ll> A(N);
	for (ll i = 0; i < N; i++)
		cin >> A[i];
	vector<ll> acc(N + 1, 0);
	for (ll i = 0; i < N; i++)
		acc[i + 1] = acc[i] + A[i];
	vector<vector<ll>> dp(N, vector<ll>(N, 1LL << 60));
	for (ll i = 0; i < N; i++)
		dp[i][i] = 0;
	for (ll k = 1; k < N; k++)
	{
		for (ll i = 0; i < N - k; i++)
		{
			ll j = i + k;
			for (ll l = i; l < j; l++)
				dp[i][j] = min(dp[i][j], dp[i][l] + dp[l + 1][j] + acc[j + 1] - acc[i]);
		}
	}
	cout << dp[0][N - 1] << endl;
}
